pkg/util/strategicpatch/patch.go

```diff
@@ -0,0 +1,454 @@
+				}
+			}
+
+			original[k] = patchV
+		}
+	}
+
+	return original, nil
+}
+
+func mergeSlice(original, patch []interface{}, elemType reflect.Type, mergeKey string) ([]interface{}, error) {
+	if len(original) == 0 && len(patch) == 0 {
+		return original, nil
+	}
+	// All the values must be of the same type, but not a list.
+	t, err := sliceElementType(original, patch)
+	if err != nil {
+		return nil, fmt.Errorf("types of list elements need to be the same, type: %s: %v",
+			elemType.Kind().String(), err)
+	}
+	if t.Kind() == reflect.Slice {
+		return nil, fmt.Errorf("not supporting merging lists of lists yet")
+	}
+	// If the elements are not maps, merge the slices of scalars.
+	if t.Kind() != reflect.Map {
+		// Maybe in the future add a "concat" mode that doesn't
+		// uniqify.
+		both := append(original, patch...)
+		return uniqifyScalars(both), nil
+	}
+
+	if mergeKey == "" {
+		return nil, fmt.Errorf("cannot merge lists without merge key for type %s", elemType.Kind().String())
+	}
+
+	// First look for any special __patch elements.
+	patchWithoutSpecialElements := []interface{}{}
+	strict := false
+	for _, v := range patch {
+		typedV := v.(map[string]interface{})
+		patchType, ok := typedV[specialKey]
+		if ok {
+			if patchType == "delete" {
+				mergeValue, ok := typedV[mergeKey]
+				if ok {
+					_, originalKey, found := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
+					if found {
+						// Delete the element at originalKey.
+						original = append(original[:originalKey], original[originalKey+1:]...)
+					}
+				} else {
+					return nil, fmt.Errorf("delete patch type with no merge key defined")
+				}
+			} else if patchType == "strict" {
+				strict = true
+				// Continue iterating through the array to prune any other __patch elements.
+			} else {
+				return nil, fmt.Errorf("unknown patch type found: %s", patchType)
+			}
+		} else {
+			patchWithoutSpecialElements = append(patchWithoutSpecialElements, v)
+		}
+	}
+
+	if strict {
+		return patchWithoutSpecialElements, nil
+	}
+
+	patch = patchWithoutSpecialElements
+
+	// Merge patch into original.
+	for _, v := range patch {
+		// Because earlier we confirmed that all the elements are maps.
+		typedV := v.(map[string]interface{})
+		mergeValue, ok := typedV[mergeKey]
+		if !ok {
+			return nil, fmt.Errorf("all list elements need the merge key %s", mergeKey)
+		}
+		// If we find a value with this merge key value in original, merge the
+		// maps. Otherwise append onto original.
+		originalMap, originalKey, found := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
+		if found {
+			var mergedMaps interface{}
+			var err error
+			// Merge into original.
+			mergedMaps, err = mergeMap(originalMap, typedV, elemType)
+			if err != nil {
+				return nil, err
+			}
+			original[originalKey] = mergedMaps
+		} else {
+			original = append(original, v)
+		}
+	}
+
+	return original, nil
+}
+
+// This panics if any element of the slice is not a map.
```

Comment:
Because it's more like an "unreachable", not "bad input". A slice of anything but all maps should never be able to reach this function. Given that if you still think it's better as an error I can convert.